nano/node/telemetry.cpp

```diff
@@ -18,469 +17,282 @@
-		}
-	},
-	nano::buffer_drop_policy::no_socket_drop);
-	// clang-format on
+	channel->send (message);
 }
 
-void nano::telemetry::channel_processed (nano::endpoint const & endpoint_a, bool error_a)
+void nano::telemetry::run_broadcasts ()
 {
-	auto it = recent_or_initial_request_telemetry_data.find (endpoint_a);
-	if (it != recent_or_initial_request_telemetry_data.end ())
+	auto telemetry = node.local_telemetry ();
+	auto peers = network.list ();
+
+	for (auto & channel : peers)
 	{
-		if (!error_a)
-		{
-			recent_or_initial_request_telemetry_data.modify (it, [] (nano::telemetry_info & telemetry_info_a) {
-				telemetry_info_a.last_response = std::chrono::steady_clock::now ();
-				telemetry_info_a.undergoing_request = false;
-			});
-		}
-		else
-		{
-			recent_or_initial_request_telemetry_data.erase (endpoint_a);
-		}
-		flush_callbacks_async (endpoint_a, error_a);
+		broadcast (channel, telemetry);
 	}
 }
 
-void nano::telemetry::flush_callbacks_async (nano::endpoint const & endpoint_a, bool error_a)
+void nano::telemetry::broadcast (std::shared_ptr<nano::transport::channel> & channel, const nano::telemetry_data & telemetry)
 {
-	// Post to thread_pool so that it's truly async and not on the calling thread (same problem as std::async otherwise)
-	workers.push_task ([endpoint_a, error_a, this_w = std::weak_ptr<nano::telemetry> (shared_from_this ())] () {
-		if (auto this_l = this_w.lock ())
-		{
-			nano::unique_lock<nano::mutex> lk (this_l->mutex);
-			while (!this_l->callbacks[endpoint_a].empty ())
-			{
-				lk.unlock ();
-				this_l->invoke_callbacks (endpoint_a, error_a);
-				lk.lock ();
-			}
-		}
-	});
+	stats.inc (nano::stat::type::telemetry, nano::stat::detail::broadcast);
+
+	nano::telemetry_ack message{ network_params.network, telemetry };
+	channel->send (message);
 }
 
-void nano::telemetry::invoke_callbacks (nano::endpoint const & endpoint_a, bool error_a)
+void nano::telemetry::cleanup ()
 {
-	std::vector<std::function<void (telemetry_data_response const &)>> callbacks_l;
-	telemetry_data_response response_data{ nano::telemetry_data (), endpoint_a, error_a };
-	{
-		// Copy data so that it can be used outside of holding the lock
-		nano::lock_guard<nano::mutex> guard (mutex);
+	debug_assert (!mutex.try_lock ());
 
-		callbacks_l = callbacks[endpoint_a];
-		auto it = recent_or_initial_request_telemetry_data.find (endpoint_a);
-		if (it != recent_or_initial_request_telemetry_data.end ())
+	nano::erase_if (telemetries, [this] (entry const & entry) {
+		// Remove if telemetry data is stale
+		if (!check_timeout (entry))
 		{
-			response_data.telemetry_data = it->data;
+			stats.inc (nano::stat::type::telemetry, nano::stat::detail::cleanup_outdated);
+			return true; // Erase
 		}
-		callbacks.erase (endpoint_a);
-	}
 
-	// Need to account for nodes which disable telemetry data in responses
-	for (auto & callback : callbacks_l)
-	{
-		callback (response_data);
-	}
-}
+		// Remove if channel that sent the telemetry is disconnected
```

Comment:
I wonder if we need to do this. Would it better to keep for a little bit and let it time out? Not sure.